app/domain/authentication/authn_jenkins/authenticator.rb

```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+require 'json'
+
+module Authentication
+  module AuthnJenkins
+    class Authenticator
+
+      Err = Errors::Authentication::AuthnJenkins
+      def initialize(env:)
+        @env = env
+      end
+
+      def webservice
+        @webservice ||= ::Authentication::Webservice.new(
+          account:            @account,
+          authenticator_name: @authenticator_name,
+          service_id:         @service_id
+        )
+      end
+
+      def http_get_request(url, username, password) 
+        begin
+          uri = URI.parse(url)
+          http = Net::HTTP.new(uri.host, uri.port)
+          http.read_timeout = 5
+          request = Net::HTTP::Get.new(uri.request_uri)
+          request.basic_auth(username, password)
+          http.request(request)
+        rescue Net::OpenTimeout => e
+          Rails.logger.error("Timeout to Jenkins host Exception #{e}. Jenkins host '#{url}' is most likely invalid. Validate this url can be accessed from the conjur server.")
+          raise Err::HostNotFound, url
+        end
+      end
+
+      def get_build_info(jenkins_url, job_path, build_number, username, password)
+        jenkins_build_endpoint = "/job/#{job_path}/#{build_number}/api/json"
+        jenkins_build_url = "#{jenkins_url}#{jenkins_build_endpoint}"
+        http_get_request(jenkins_build_url, username, password)
+      end
+
+      def build_running?(response)
+        body = response.body
+        status_code = response.code
+
+        Rails.logger.debug("Jenkins Build Response: #{body}")
+        if status_code != '200'
+          raise Err::BuildInfoError.new(jenkinsBuildURL, "#{status_code} - #{body}")
+        end
+
+        JSON.parse(body)['building']
+      end
+
+      def jenkins_public_key(response)
+        public_key = response['X-Instance-Identity']
+
+        public_key = "-----BEGIN PUBLIC KEY-----\n#{public_key}\n-----END PUBLIC KEY-----"
+        Rails.logger.debug("Returned public key: #{public_key}")
+        OpenSSL::PKey::RSA.new(public_key)
+      end
+
+      def parse_metadata(username, password)
```

Comment:
Although the key information comes in through the `password` attribute, what do you think about changing it to something like `message_body` to make it more clear that it's a JSON string?

Reply:
`parse_metadata` isn't a super descriptive. Is there something more descriptive we can call it?  Maybe `extract_job_name_path_build_signature`?  I also wonder if there is a better way to handle the extraction that's happening inside this method.

Reply:
Not sure if it's a net win, but it could be abstracted out into another class:
```ruby
class JenkinsLoad
  attr_reader :build_number, :signature, :job_name, :job_path

  def initialize(raw_body, username)
    body = JSON.parse(raw_body)
    @build_number = body['buildNumber']
    @signature = Base64.decode64(body['signature'])
    @job_name = if json_body.key?("jobProperty_hostPrefix")
      username.sub("host/#{job_property_host_prefix}/", "")
    else
      username.sub("host/", "")
    end
    @job_path = @job_name.split("/").join("/job/")
  end
end
```
Which would allow us to do something like:
```ruby
load = JenkinsLoad.new(password)
load.build_number
...
```

Reply:
Added a JenkinsLoad class. 
https://github.com/AndrewCopeland/conjur/commit/6a6526c2d914cc0d24e48569fcd9b61312c46577
And removed the parse_metadata:
https://github.com/AndrewCopeland/conjur/commit/91e297962396780b27521b9b9b0b8f73773449a8